drivers/display/display_st7701.c

```diff
@@ -0,0 +1,457 @@
+	uint16_t yres;
+	uint8_t dsi_pixel_format;
+	enum display_pixel_format pixel_format;
+	enum display_orientation orientation;
+};
+
+static inline int st7701_dcs_write(const struct device *dev, uint8_t cmd, const void *buf,
+		size_t len)
+{
+	const struct st7701_config *cfg = dev->config;
+	int ret;
+
+	ret = mipi_dsi_dcs_write(cfg->mipi_dsi, cfg->channel, cmd, buf, len);
+	if (ret < 0) {
+		LOG_ERR("DCS 0x%x write failed! (%d)", cmd, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int st7701_short_write_1p(const struct device *dev, uint8_t cmd, uint8_t val)
+{
+	const struct st7701_config *cfg = dev->config;
+	int ret;
+	uint8_t buf[] = {cmd, val};
+
+	ret = mipi_dsi_generic_write(cfg->mipi_dsi, cfg->channel, buf, sizeof(val));
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+
+static int st7701_generic_write(const struct device *dev, const void *buf, size_t len)
+{
+	const struct st7701_config *cfg = dev->config;
+	int ret;
+
+	ret = mipi_dsi_generic_write(cfg->mipi_dsi, cfg->channel, buf, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int st7701_check_id(const struct device *dev)
+{
+	const struct st7701_config *cfg = dev->config;
+	uint32_t id = 0;
+	int ret;
+
+	ret = mipi_dsi_dcs_read(cfg->mipi_dsi, cfg->channel, ST7701_CMD_ID1, &id, sizeof(id));
+	if (ret != sizeof(id)) {
+		LOG_ERR("Read panel ID failed! (%d)", ret);
+		return -EIO;
+	}
+
+	if (id != 0xFF) {
+		LOG_ERR("ID 0x%x (should 0x%x)", id, 0xFF);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int st7701_configure(const struct device *dev)
+{
+	struct st7701_data *data = dev->data;
+	uint8_t buf[4];
+	int ret;
+
+	const uint8_t Display_Control_0[] = {0xFF, 0x77, 0x01, 0x00, 0x00, 0x10};
+	const uint8_t Display_Control_1[] = {0xC0, 0x63, 0x00};
+	const uint8_t Display_Control_2[] = {0xC1, 0x11, 0x02};
+	const uint8_t Display_Control_3[] = {0xC2, 0x01, 0x08};
+	const uint8_t Display_Control_4[] = {0xCC, 0x18};
+
+	st7701_generic_write(dev, Display_Control_0, sizeof(Display_Control_0));
+	st7701_generic_write(dev, Display_Control_1, sizeof(Display_Control_1));
+	st7701_generic_write(dev, Display_Control_2, sizeof(Display_Control_2));
+	st7701_generic_write(dev, Display_Control_3, sizeof(Display_Control_3));
+	st7701_generic_write(dev, Display_Control_4, sizeof(Display_Control_4));
+
+	/* Gamma Cluster Setting */
+	const uint8_t _B0[] = { 0xB0, 0x40, 0xc9, 0x91, 0x0d,
+							0x12, 0x07, 0x02, 0x09, 0x09,
+							0x1f, 0x04, 0x50, 0x0f, 0xe4,
+							0x29, 0xdf };
+
+	const uint8_t _B1[] = { 0xB1, 0x40, 0xcb, 0xd0, 0x11,
+							0x92, 0x07, 0x00, 0x08, 0x07,
+							0x1c, 0x06, 0x53, 0x12, 0x63,
+							0xeb, 0xdf };
+
+	st7701_generic_write(dev, _B0, sizeof(_B0));
+	st7701_generic_write(dev, _B1, sizeof(_B1));
+
+	/* Initial power control registers */
+	const uint8_t _FF1[] = {DSI_CMD2BKX_SEL, 0x77, 0x01, 0x00, 0x00, DSI_CMD2BK1_SEL};
+
+	st7701_generic_write(dev, _FF1, sizeof(_FF1));
+
+	st7701_short_write_1p(dev, DSI_CMD2_BK1_VRHS, 0x65);
```

Comment:
I get that error checking every one of these is tedious but some logging in case those fails would be good. Maybe inside the method itself?

Reply:
I've added logging to those functions, which also resolves some of the other comments.